File: jvm/doris-ducklake/test/src/dev/brikk/ducklake/doris/plugin/DuckLakeConnectorCapabilitiesTest.java
package dev.brikk.ducklake.doris.plugin;

import java.util.EnumSet;
import java.util.Map;
import java.util.Optional;

import dev.brikk.ducklake.doris.plugin.cache.FakeConnectorContext;

import org.apache.doris.connector.api.Connector;
import org.apache.doris.connector.api.ConnectorCapability;
import org.apache.doris.connector.api.timetravel.ConnectorMvccSnapshot;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * Pins the v1 capability declaration on {@link DuckLakeConnector} and the
 * matching codec wiring on {@link DuckLakeConnectorProvider}. Sanity-check §4
 * is the load-bearing reference. Adding or removing entries from the EnumSet
 * is a deliberate roadmap step (see {@code ducklake-doris-todo.md}) — bump
 * the assertion in the same PR that flips a feature on.
 */
class DuckLakeConnectorCapabilitiesTest {

    @Test
    void declaresOnlyV1Capabilities() {
        Connector connector = new DuckLakeConnector(
                Map.of(),
                new FakeConnectorContext("dl", 1L));
        assertThat(connector.getCapabilities())
                .containsExactlyInAnyOrderElementsOf(EnumSet.of(
                        ConnectorCapability.SUPPORTS_MVCC_SNAPSHOT,
                        ConnectorCapability.SUPPORTS_POSITION_DELETE,
                        ConnectorCapability.SUPPORTS_TIME_TRAVEL,
                        ConnectorCapability.SUPPORTS_PARTITION_PRUNING,
                        ConnectorCapability.SUPPORTS_STATISTICS));
    }

    @Test
    void pushdownCapabilitiesStayOffUntilApplyMethodsLand() {
        // Declaring without implementing crashes the planner — keep these
        // off until Step 6 of the roadmap implements the apply* hooks.
        Connector connector = new DuckLakeConnector(
                Map.of(),
                new FakeConnectorContext("dl", 1L));
        assertThat(connector.getCapabilities())
                .doesNotContain(
                        ConnectorCapability.SUPPORTS_FILTER_PUSHDOWN,
                        ConnectorCapability.SUPPORTS_PROJECTION_PUSHDOWN,
                        ConnectorCapability.SUPPORTS_LIMIT_PUSHDOWN);
    }

    @Test
    void providerSuppliesMvccSnapshotCodec() {
        DuckLakeConnectorProvider provider = new DuckLakeConnectorProvider();
        Optional<ConnectorMvccSnapshot.Codec> codec = provider.getMvccSnapshotCodec();
        assertThat(codec).isPresent();
        assertThat(codec.get()).isInstanceOf(DuckLakeConnectorMvccSnapshot.Codec.class);
    }
}

File: jvm/doris-ducklake/test/src/dev/brikk/ducklake/doris/plugin/DuckLakeConnectorCapabilitiesTest.kt
package dev.brikk.ducklake.doris.plugin

import java.util.EnumSet

import dev.brikk.ducklake.doris.plugin.cache.FakeConnectorContext

import org.apache.doris.connector.api.Connector
import org.apache.doris.connector.api.ConnectorCapability
import org.junit.jupiter.api.Test

import org.assertj.core.api.Assertions.assertThat

/**
 * Pins the capability declaration on [DuckLakeConnector] against the P6
 * iceberg-cutover SPI (`branch-catalog-spi` @ 8b391c7), where the
 * capability-unification gutted the declarative enum: pushdown / insert /
 * create-table / time-travel / statistics constants were deleted (fe-core
 * never consumed them — pushdown is attempted unconditionally, INSERT
 * admission is `supportedWriteOperations()`, DDL routes through
 * `ConnectorTableOps`). `SUPPORTS_MVCC_SNAPSHOT` is the surviving flag that
 * matters: it gates MVCC table creation (snapshot pinning + time travel).
 *
 * Adding entries from the new opt-in set (`SUPPORTS_SHOW_CREATE_DDL`,
 * `SUPPORTS_TOPN_LAZY_MATERIALIZE`, `SUPPORTS_NESTED_COLUMN_PRUNE`,
 * `SUPPORTS_VIEW`, `SUPPORTS_COLUMN_AUTO_ANALYZE`,
 * `SUPPORTS_METADATA_PRELOAD`) is a deliberate roadmap step (see
 * `dev-docs/TODO-read.md` + `REPORT-doris-p6-iceberg-spi-cutover.md`) —
 * bump the assertion in the same PR that flips a feature on.
 */
internal class DuckLakeConnectorCapabilitiesTest {

    @Test
    fun declaresOnlyMvccSnapshot() {
        val connector: Connector = DuckLakeConnector(
            emptyMap(),
            FakeConnectorContext("dl", 1L),
        )
        assertThat(connector.capabilities)
            .containsExactlyInAnyOrderElementsOf(
                EnumSet.of(
                    ConnectorCapability.SUPPORTS_MVCC_SNAPSHOT,
                ),
            )
    }

    @Test
    fun newOptInCapabilitiesStayOffUntilTheirPrerequisitesLand() {
        // Each of these has behavioral prerequisites (field-id stamping for
        // nested prune, credential-free properties for SHOW CREATE, ...);
        // declaring one without the matching implementation misplans queries.
        val connector: Connector = DuckLakeConnector(
            emptyMap(),
            FakeConnectorContext("dl", 1L),
        )
        assertThat(connector.capabilities)
            .doesNotContain(
                ConnectorCapability.SUPPORTS_SHOW_CREATE_DDL,
                ConnectorCapability.SUPPORTS_TOPN_LAZY_MATERIALIZE,
                ConnectorCapability.SUPPORTS_NESTED_COLUMN_PRUNE,
                ConnectorCapability.SUPPORTS_VIEW,
                ConnectorCapability.SUPPORTS_COLUMN_AUTO_ANALYZE,
                ConnectorCapability.SUPPORTS_METADATA_PRELOAD,
            )
    }
}
